src/history.rs

```diff
@@ -403,21 +364,42 @@ struct HistoryImpl {
     old_item_offsets: Vec<usize>,
 }
 
-/// If set, we gave up on file locking because it took too long.
-/// Note this is shared among all history instances.
-static ABANDONED_LOCKING: RelaxedAtomicBool = RelaxedAtomicBool::new(false);
-
 impl HistoryImpl {
+    /// Returns the canonical path for the history file, or an error.
+    /// An error is returned for an empty name (which indicates private mode).
+    /// An error is returned if obtaining the data directory failed.
+    /// Because the `path_get_data` function does not return error information,
+    /// we cannot provide more detail about the reason for the failure here.
+    fn history_file_path(&self) -> std::io::Result<WString> {
+        if self.name.is_empty() {
+            return Err(std::io::Error::new(
+            std::io::ErrorKind::Other,
+            "Obtaining the history filename is not possible when the name is empty (private mode)",
+        ));
+        }
+
+        let Some(mut path) = path_get_data() else {
+            return Err(std::io::Error::new(std::io::ErrorKind::NotFound, "Error obtaining data directory. This is a manually constructed error which does not indicate why this happened."));
+        };
+
+        path.push('/');
+        path.push_utfstr(&self.name);
+        path.push_utfstr(L!("_history"));
+        if let Some(canonicalized_path) = wrealpath(&path) {
+            Ok(canonicalized_path)
+        } else {
+            Err(std::io::Error::new(
+                std::io::ErrorKind::Other,
+                format!("wrealpath failed to produce a canonical version of '{path}'."),
+            ))
+        }
+    }
+
     /// Add a new history item to the end. If `pending` is set, the item will not be returned by
     /// `item_at_index()` until a call to `resolve_pending()`. Pending items are tracked with an
     /// offset into the array of new items, so adding a non-pending item has the effect of resolving
     /// all pending items.
-    fn add(
-        &mut self,
-        item: HistoryItem,
-        pending: bool, /*=false*/
-        do_save: bool, /*=true*/
-    ) {
+    fn add(&mut self, item: HistoryItem, pending: bool, do_save: bool) {
```

Comment:
yeah this is a holdover from C++ default arguments (to ease porting) but there's no such thing in Rust